### PR TITLE
DBS pypi packaging

### PR DESCRIPTION
**Pull request by goughes**
This PR is related to #540. Do not merge yet as it is a work in progress.  I have added requirements.txt and setup scripts that can be used to upload the DBS client and Pycurl client to pypi.  These are currently available on pypi and can be installed via pip.

Some docs: https://github.com/dmwm/WMCore/wiki/DMWM-Packaging-with-PyPi
PyPI page for DMWM: https://pypi.org/user/cms-oc-dmwm/

Regarding versioning, pypi packages need to follow PEP 440: https://www.python.org/dev/peps/pep-0440/
DBS tags do not follow this scheme, so we should discuss if/how they can be changed.

### CODE REVIEW

**Comment by goughes** [comment]
Spoke with Rucio devs regarding how they build packages and they pointed me to this: https://github.com/rucio/rucio/blob/master/tools/build_sdist_and_upload.sh
I added a similar script for DBS.

That leaves this question.  What other software from the DBS repo needs to be available in pypi? dbs-web? dbs-migration?

**Comment by yuyiguo** [comment]
Only DBS client and pycurl client.

**Comment by amaltaro** [comment]
@goughes Erik, I'm wondering whether it looks cleaner to simply add that setup_functions code to each of those setup_* scripts? Just so we keep all code related to building pypi packages in the same place.

Maybe we can "blindly" open that file, if requirements cannot be found an exception will be raised and the packaging process will be stopped (that will save us a few lines in the code :))

**Comment by yuyiguo** [comment]
@goughes Erik, Could you please add instructions on how to install DBS client via pypi ? 
DBS Client currently deployed with Crab. The instruction https://twiki.cern.ch/twiki/bin/viewauth/CMS/DBS3APIInstructions points how to get DBS Client through crab. It would be nice that you could add the instructions via pypi there.
Thanks a lot.

**Comment by goughes** [comment]
@amaltaro You are right.  I have removed setup_functions so all the code needed to build a package is in a the individual setup files.

**Comment by goughes** [comment]
@amaltaro @yuyiguo I updated the instructions here for how to build and upload new PyPI package and install DBS client via PyPI.  Can you take a look?  DM me on Slack if you want the password for uploading.
https://github.com/dmwm/WMCore/wiki/DMWM-Packaging-with-PyPi#packaging-dbs-applications

**Comment by yuyiguo** [comment]
@goughes Erik, It seems this pypi installing DBS is really useful to end users. Currently DBS client is distributed with crab client which is not all the DBS users need.  I tried to install DBS client on my laptop, but I did not have virtualenv . The end users' own machines  usually may not have virtualenv. If you could add instruction on installing virtualenv and remove the dependency on CMS library, then we could advertise this.

**Comment by amaltaro** [comment]
@goughes @yuyiguo I added a couple of commands for installing pip and virtualenv, then after installing libcurl-devel I managed to install dbs-client. However, calling the dbsApis fail with the pycurl.error error you already reported. Trying the cvmfs tricky didn't help either.

However, if I instead use a pair of cert/key service certificate files, then it works just fine.

**Comment by yuyiguo** [comment]
Thanks Alan for updating the instruction. Could you please also add the instructions on how to use cert/key files to make it really works?

**Comment by amaltaro** [comment]
For that you would need to have access to a service/robot certificate. So I don't know how we can make it work for a "mortal" user.

**Comment by goughes** [comment]
Seems like this might be related to some special sauce CMS puts into curl... anyone know who I could ask for some additional info?
```
(dbs7) [egough@lxplus031 ~]$ python dbs-test.py 
Traceback (most recent call last):
  File "dbs-test.py", line 3, in <module>
    ds = dbs.listDatasets(dataset = '/MinBias/Summer*/*')
  File "/afs/cern.ch/user/e/egough/dbs7/lib/python2.7/site-packages/dbs/apis/dbsClient.py", line 758, in listDatasets
    return self.__callServer("datasets", params=kwargs)
  File "/afs/cern.ch/user/e/egough/dbs7/lib/python2.7/site-packages/dbs/apis/dbsClient.py", line 197, in __callServer
    self.http_response = method_func(self.url, method, params, data, request_headers)
  File "/afs/cern.ch/user/e/egough/dbs7/lib/python2.7/site-packages/RestClient/RestApi.py", line 34, in get
    return http_request(self._curl)
  File "/afs/cern.ch/user/e/egough/dbs7/lib/python2.7/site-packages/RestClient/RequestHandling/HTTPRequest.py", line 57, in __call__
    curl_object.perform()
pycurl.error: (35, 'Peer does not recognize and trust the CA that issued your certificate.')
(dbs7) [egough@lxplus031 ~]$ export LD_LIBRARY_PATH=/cvmfs/cms.cern.ch/slc7_amd64_gcc630/external/curl/7.59.0/lib
(dbs7) [egough@lxplus031 ~]$ python dbs-test.py 
[{'dataset': '/MinBias/Summer09-MC_31X_V3-v1_HLT_AlCa1E31-v1/RAW'}, {'dataset': '/MinBias/Summer09-MC_31X_V3-v1_HLT_AlCa8E29-v1/RAW'}, {'dataset': '/MinBias/Summer09-MC_31X_V3_CS_DiJetAve15U-v1/USER'}, {'dataset': '/MinBias/Summer09-MC_31X_V3_7TeV_CS_DiJetAve15U-v1/USER'}, {'dataset': '/MinBias/Summer11-DESIGN42_V11_428_SLHC1-v1/GEN-SIM'}, {'dataset': '/MinBias/Summer11-DESIGN42_V11_428_SLHChcal-v1/GEN-SIM'}]
```

**Comment by belforte** [comment]
example (I used one curl line from CRAB op/debug twiki).
In all cases after voms-proxy-init -voms CMS and haveing defined X509_USER_PROXY

* plain lxplus:
```
belforte@lxplus091/~> curl -X GET 'https://cmsweb.cern.ch/crabcache/info?subresource=userinfo&username=mmascher' --key $X509_USER_PROXY --cert $X509_USER_PROXY -k
curl: (35) SSL connect error
belforte@lxplus091/~> 
```

* lxplus + CMSSW
```
belforte@lxplus091/belforte> curl -X GET 'https://cmsweb.cern.ch/crabcache/info?subresource=userinfo&username=mmascher' --key $X509_USER_PROXY --cert $X509_USER_PROXY -k
{"result": [
 {"file_list": [], "used_space": [0]}
]}
belforte@lxplus091/belforte> 
```

* ubuntu 18.04 w/o CMSSW
```
LapSB:CRABServer$ curl -X GET 'https://cmsweb.cern.ch/crabcache/info?subresource=userinfo&username=mmascher' --key $X509_USER_PROXY --cert $X509_USER_PROXY -k
{"result": [
 {"file_list": [], "used_space": [0]}
]}
LapSB:CRABServer$ 

```

**Comment by amaltaro** [comment]
Thanks, Stefano. Yes, I think you did, but I let Erik discuss these details with you since he's the one really investigating this issue.

I do have a few comments though:
>  One needs the version from CMS.

which is available either in the CMSSW environment or in our CMS COMP one.
Then looking at cmsdist/comp_gcc630, I see dbs3-client depends on dbs3-pycurl-client, which depends on curl, and the curl spec is built with SSL support. So I still fail to see why it doesn't work.

**Comment by vkuznet** [comment]
I think it is missing environment settings in pypi installation. In CMSSW or CMS COMP every package has `etc/profile.d/init.sh` and `etc/profile.d/dependencies.sh` files which sets PYTHONPATH and LD_LIBRARY_PATH properly. Since now you're installing via pypi these settings are missing in your setup. What you see is that by running `python dbs-test.py` the latter script will load ssl and take system curl shared library. As you demonstrated when you setup LD_LIBRARY_PATH to point to proper curl library the behavior is fixed. But unfortunately it is not proper way to do it due to recursive chain of dependencies which we set through `etc/profile.d/init.sh`. You need to source this script which will go through all packages and setup full PYTHONPATH and LD_LIBRARY_PATH. I doubt that there is a mechanism in pypi installation to load this script by default.

**Comment by vkuznet** [comment]
But I also think that CMSSW folks (David Lange, Shahzad) resolved somehow this issue since they should have identical problem installing and running CMSSW packages in pypi/virtualenv. I would suggest to talk to them about it.

**Comment by belforte** [comment]
Well.. what is exactly the goal here ?
Indeed CMS_COMP provides the proper libcurl and if the goal is to use pip to run DBS Client in the CMS_COMP environment w/o esplictely sourcing the CMS_COMP setup first,
you need to tackle what Valenting pointed out.

Are you looking into getting rid of CMS_COMP for WMA ? If not, just source it etc/profile.d/initl.sh and go. 

I thought you wanted to make it possible to run DBS Client "on any SL6/7 box', in that case the
first question is 'where to get libcurl'.

But if the goal is to make it easy to use for CMS users, we can assume that they have the CMSSW environment defined and IIUC simply use pip to add DBS client to it w/o bothering with a full CMS_COMP distribution (which would be the current proper way).
As Yuyi pointed out, getting DBS Client python files extracted from gtihub and sticked inside CRAB Client CVMFS directories is not so nice, while I do expect a technical problem (just add a few python files to the ones for CRAB), for sure it makes version control difficult and is somehow un-natural since
CRAB Client does not talk to DBS.

**Comment by goughes** [comment]
Thanks for the comments everyone. 

The original goal of this PR was to get DBS client and pycurl client into pypi so they could be pulled as dependencies when installing WMAgent via pip.  I think we have met that goal as the curl issue with user certs or proxies I describe above will not exist in that scenario.

Giving users to install a working dbs-client on **any** client would be quite tricky and is probably out of the scope of this PR.  At a minimum, I can put some instructions for installing dbs-client via pip on lxplus7 with the CMSSW environment.

**Comment by belforte** [comment]
Tahnks @goughes 
When you have instructions for users install on top of CMSSW, it may be good to add to existing twiki:
https://twiki.cern.ch/twiki/bin/view/CMS/DBS3APIInstructions

**Comment by amaltaro** [comment]
@goughes can you please follow the instructions you wrote down and upload the latest dbs-client to Pypi (3.7.8)? If everything goes well, then I'd suggest we merge this PR and follow up the curl/architecture limitations in another ticket. As Erik mentioned, the main goal was to get WMAgent deployable via Pypi.

@yuyiguo if you agree, then let's wait for Erik's confirmation and we can move on with this.

**Comment by yuyiguo** [comment]
Alan,
That is fine with me. But what I understood was that DBS client was used as guinea pig to deploy DM code via Pypi.
Cheers,
Yuyi

**Comment by goughes** [comment]
I uploaded the 3.7.8 version of dbs client and pycurl client to PyPi.  I was also able to install and make a successful query to the DBS api using the updated instructions I put here https://github.com/dmwm/WMCore/wiki/DMWM-Packaging-with-PyPi#installing-dbs-client-and-pycurl-client-via-pypi-on-lxplus7

I did not have any success mixing a CMSSW env with a python virtualenv, so the instructions use the `--user` option of pip to install the package to the .local dir in the user's home directory.

**Comment by belforte** [comment]
isn't this what anybody w/o root access will get by default anyhow ?

> I did not have any success mixing a CMSSW env with a python virtualenv, so the instructions use the `--user` option of pip to install the package to the .local dir in the user's home directory.

**Comment by goughes** [comment]
If you don't specify --user it tries to install into the system python directories and you get an error.

`OSError: [Errno 13] Permission denied: '/usr/lib/python2.7/site-packages/RestClient'`

**Comment by belforte** [comment]
thanks @goughes and sorry for noise, I was using confused memories.

**Comment by yuyiguo** [comment]
@amaltaro I don't have anything on the pypi package. We depend on the author to write the package summary and document as well as unit tests, deployment tests if they are applicable. If everything is done, feel free to commit.

**Comment by amaltaro** [comment]
Well, the author of DBS packages/services is you ;-) Erik simply did the upload and the setup to get this available.
Anyhow, I'm not asking you to give us anything now, this is for the future... Merging this one would be a good start IMO.

**Comment by yuyiguo** [comment]
@amaltaro I did not write the pypi part. Anyone contribute to any package in CMS is more than welcome as I understood. However, Contributing to a package is more than writeing the code, it includes document, and tests. I hope we are in the same page.

**Comment by belforte** [comment]
It may be off topic, but I find ti REALLY GREAT that now I can run dbs client on my Ubuntu laptop !
(latest Ubuntu has proper libcurl out of the box).

I am interested also in wmcore, I see that pip has 1.1.1rc7, do you plan to make and keep that up to date ?

**Comment by vkuznet** [comment]
Stefano,
as I reported on meeting, we should be careful with installations until
proper license will be in place.
Best,
Valentin.

**Comment by belforte** [comment]
fine with being careful, i am simply a bit lost as to what the goal
is and how this impacts CRAB

**Comment by yuyiguo** [comment]
The same question here. What is the goal of all this? What kind license do I need ?
But it is still a good news that Stefano can run it out of box w/o CMS specific software decency.  Can we improve DBS pypi package to get the newest libcurl?
Thanks,
Yuyi

**Comment by vkuznet** [comment]
Yuyi, on a WMA meeting we discussed this. Original goal was to simplify WMA
deployment and not a public release of DBS client.

Regarding license, you need to understand that whatever you release to public
domain (in this case pypi) can be used for different purposes. All CMS software
is open source, but according to this note:

https://indico.cern.ch/event/791214/attachments/1783962/2906417/CBnewsletter-CB122-February-2019.pdf

the collaboration board discuss a possibility to release all CMS software
under apache 2 license instead of GPL one (may be it is default one), see item 7.

I suggested on a meeting that our release of software to public should be
adopted with certain license and I asked Alan to clarify what it should be.

As developers we should be aware of how our software is distributed
to avoid possible issues. That's why I made my suggestions that some
license should be added to pypi packages.

Best,
Valentin.

**Comment by yuyiguo** [comment]
Thanks for clarifying things. There was some miss information I got before.
Yuyi

**Comment by amaltaro** [comment]
> Yuyi, on a WMA meeting we discussed this. Original goal was to simplify WMA deployment and not a public release of DBS client.

if you want to simplify the deployment procedure by using Pypi packages instead of RPM, how can you possibly not make it available to the general public?!?

As I said before a couple of times, and so did Erik, the main goal of these developments is to be able to deploy WMAgent via Pypi. If someone else installs it, then it's on their own.

About the license, sure, let's discuss it and get it properly done with the CMS packages. However, what's out there is already there and there is nothing that we can do but discussing the correct license and making the necessary changes. I don't think the license thing is a show stopper to proceed with this PR.

Yuyi, if you feel there is anything else missing here, please let us know such that we can move forward.

**Comment by yuyiguo** [comment]
Alan,
I did not know about the goal of making DBS client in pypi or the license issue involved so I asked. Valentin's reply made sense. What I said stopped you moving forward? If so, I might be miss understood.

Stefano's  comment regarding the newest libcurl could be something to make the package to be used out of box. So I ask if it can be improved. What you think?
Thanks,
Yuyi

**Comment by yuyiguo** [comment]
@belforte  Hi Stefano, Could you let us know the version of libcurl in your new Ubuntu machine?
Thanks,
Yuyi

**Comment by belforte** [comment]
I wish I knew how tpo answer
I had to install (ana)conda to get xroot client and now it overrode things, so I have:

LapSB:~$ which curl
/home/belforte/mybin/anaconda2/bin/curl
LapSB:~$ curl --version
curl 7.63.0 (x86_64-conda_cos6-linux-gnu) libcurl/7.63.0 OpenSSL/1.0.2p zlib/1.2.11 libssh2/1.8.0
Release-Date: 2018-12-12
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy
LapSB:~$

while in standard place I have:
/usr/lib/x86_64-linux-gnu/libcurl.so.4
/usr/lib/x86_64-linux-gnu/libcurl.so.4.5.0

and if you want some details about it, I do not now how to get them
.

I do not know how to go back to non-conda shell.
Anyhow it was just ubuntu 18.04 LTS and curl have been always working for me.

**Comment by vkuznet** [comment]
Stefano,
to go out from conda environment you should deactivate it:
conda deactivate

Then Ubuntu is debian based Linux distro, so it uses apt tools.
You can list installed packages as easy as following:
apt list --installed

Best,
Valentin.

**Comment by belforte** [comment]
thanks Valentin , so here's my setup , hope it is what Yuyi wants:
LapSB:~$ which curl
/usr/bin/curl
LapSB:~$ curl --version
curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.0g zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
Release-Date: 2018-01-24
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
LapSB:~$ 
LapSB:WMCore$ apt list --installed|grep curl

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

curl/bionic-updates,bionic-security,now 7.58.0-2ubuntu3.6 amd64 [installed,automatic]
libcurl3-gnutls/bionic-updates,bionic-security,now 7.58.0-2ubuntu3.6 amd64 [installed]
libcurl4/bionic-updates,bionic-security,now 7.58.0-2ubuntu3.6 amd64 [installed,automatic]
LapSB:WMCore$ which curl
/home/belforte/mybin/anaconda2/bin/curl
LapSB:WMCore$

**Comment by yuyiguo** [comment]
Thanks a lot , @belforte @vkuznet for finding the libcurl version.
@goughes is possible to have a dependency on the newer version of libcurl?  So we don't have to reply on cms software. Please refer to Stefano's massage for the version.

**Comment by vkuznet** [comment]
Yuyi, we can't make dependency on libcurl since it is not python package. The
dependency of python packages on pypi defines other python packages. If your
python package relies on C-library (in this case libcurl) I doubt that you
can make it since libcurl is not part of pypi repository.

**Comment by yuyiguo** [comment]
Valentin,
I see what you meant. I used a wrong word "dependency".  Let me try again and use an example.
I'd like to pip install lxml, but lxml requires libxslt  and libxml2.  My system has only libxslt with the correct version, libxml2 was too old. So I had to use brew to install the newer libxml2 first. The requirement was in the lxml doc. 

If we can figure out which version of libcurl works, then we can make people to aware the requirement of DBS client.

I hope I made things clear.
Yuyi

**Comment by bbockelm** [comment]
Hi,

The way RHEL compiles `libcurl`, it doesn't work with grid proxies out-of-the-box.  CMS uses a special configuration.

So, it's a far trickier job than just upgrading a version.

That said, if you open a separate ticket on the topic, I can probably outline what needs to be done in WMCore-land to make grid proxies work with `libcurl`.

Brian

**Comment by yuyiguo** [comment]
Got it. Thanks Brain.

**Comment by amaltaro** [comment]
Just to x-reference, Yuyi created a ticket to address the libcurl issue here:
https://github.com/dmwm/WMCore/issues/9036

@goughes you might want to squash your commits.

Thanks everyone for helping out here.

**Comment by goughes** [comment]
Added some lines to include the license files to the packages that get uploaded to pypi.  This depends on #596 being merged, but I have tested locally.

**Comment by yuyiguo** [comment]
@goughes This is a copy/paste from the other PR.
Erik,  It makes sense to me that the license files and pypi setup files on the the top DBS directory as Erik explained. I'd like to see requirements.dbs_client.txt, requirements.pycurl_client.txt, setup_dbs_client.py and setup_pycurl_client.py in their own directories(client and pycurl client). As soon as these changes made, I will commit them.
Sorry Erik, I missed your comment on license files.
THanks,
Yuyi

**Comment by goughes** [comment]
@yuyiguo I made the changes you requested.  Can you take a look?

**Comment by yuyiguo** [comment]
@goughes Thank you for making the changing. We are having very serious DBS production problem right now. I will look into this when I have time.

**Comment by amaltaro** [comment]
Thanks, Yuyi!
